File: app/src/main/java/org/ametro/ui/widgets/MapTopPanelWidget.java
package org.ametro.ui.widgets;


import android.animation.Animator;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import org.ametro.R;
import org.ametro.app.Constants;

public class MapTopPanelWidget implements Animator.AnimatorListener {

    private final View view;
    private final TextView textView;
    private String text;

    private final Runnable hideAnimation = new Runnable() {
        @Override
        public void run() {
            view.animate().setDuration(Constants.ANIMATION_DURATION).setListener(MapTopPanelWidget.this).translationY(-view.getHeight());
        }
    };

    private final Runnable showAnimation = new Runnable() {
        @Override
        public void run() {
            view.setVisibility(View.VISIBLE);
            textView.setText(text);
            view.animate().setDuration(Constants.ANIMATION_DURATION).setListener(MapTopPanelWidget.this).translationY(0);
        }
    };

    private Runnable actionOnEndAnimation;
    private boolean visible;
    private boolean firstTime;

    public MapTopPanelWidget(ViewGroup view) {
        this.view = view;
        this.textView = (TextView)view.findViewById(R.id.message);
        this.visible = false;
        this.firstTime = true;
    }

    public void show(final String newText) {
        if (visible && (text != null && newText!=null && text.equals(newText))) {
            return;
        }

        text = newText;

        if (!visible && !firstTime) {
            visible = true;
            showAnimation.run();
            return;
        }

        visible = true;
        firstTime = false;
        actionOnEndAnimation = showAnimation;
        hideAnimation.run();
    }

    public void hide() {
        if (!visible) {
            return;
        }
        visible = false;
        hideAnimation.run();
    }

    @Override
    public void onAnimationStart(Animator animation) {

    }

    @Override
    public void onAnimationEnd(Animator animation) {
        if (!visible)
            view.setVisibility(View.INVISIBLE);

        if (actionOnEndAnimation != null) {
            actionOnEndAnimation.run();
            actionOnEndAnimation = null;
        }
    }

    @Override
    public void onAnimationCancel(Animator animation) {

    }

    @Override
    public void onAnimationRepeat(Animator animation) {

    }
}

File: app/src/main/java/io/github/romangolovanov/apps/ametro/ui/widgets/MapTopPanelWidget.kt
package io.github.romangolovanov.apps.ametro.ui.widgets

import android.animation.Animator
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import io.github.romangolovanov.apps.ametro.R
import io.github.romangolovanov.apps.ametro.app.Constants

class MapTopPanelWidget(viewGroup: ViewGroup) : Animator.AnimatorListener {

    private val view: View = viewGroup
    private val textView: TextView = viewGroup.findViewById(R.id.message)
    private var text: String? = null

    private val hideAnimation = Runnable {
        view.animate()
            .setDuration(Constants.ANIMATION_DURATION)
            .setListener(this@MapTopPanelWidget)
            .translationY(-view.height.toFloat())
    }

    private val showAnimation = Runnable {
        view.visibility = View.VISIBLE
        textView.text = text
        view.animate()
            .setDuration(Constants.ANIMATION_DURATION)
            .setListener(this@MapTopPanelWidget)
            .translationY(0f)
    }

    private var actionOnEndAnimation: Runnable? = null
    private var visible = false
    private var firstTime = true

    fun show(newText: String?) {
        if (visible && text != null && newText != null && text == newText) return

        text = newText

        if (!visible && !firstTime) {
            visible = true
            showAnimation.run()
            return
        }

        visible = true
        firstTime = false
        actionOnEndAnimation = showAnimation
        hideAnimation.run()
    }

    fun hide() {
        if (!visible) return
        visible = false
        hideAnimation.run()
    }

    override fun onAnimationStart(animation: Animator) {}

    override fun onAnimationEnd(animation: Animator) {
        if (!visible) view.visibility = View.INVISIBLE
        actionOnEndAnimation?.run()
        actionOnEndAnimation = null
    }

    override fun onAnimationCancel(animation: Animator) {}

    override fun onAnimationRepeat(animation: Animator) {}
}
